### DISCUSSION TITLE
how to use two await in middleware?

### DISCUSSION THREAD

**Discussion by panla** [fastapi/fastapi#6724] [Questions] [answered]
### how to use two await in middleware?

```python
from fastapi.logger import logger
app.logger = logger
```

```python
@app.middleware('http')
async def after_request(request: Request, call_next):
    params = await request.json()
    request.app.logger.info(params)
    response = await call_next(request)
    return response
```

if I use two await in one middleare, this is block and it can not continue code

how to done it?

**Comment by juntatalor** [comment, marked as answer]
@panla Hello!
Could you please provide a more complex example of blocking middleware? I have just tested your code and it works fine (because json() method does not re-read body, it uses internal `_json` field for caching)
My working code:

```python
from fastapi.logger import logger
from fastapi import FastAPI, Request
import logging

logging.basicConfig(level=logging.INFO)

app = FastAPI()
app.logger = logger

@app.middleware('http')
async def after_request(request: Request, call_next):
    params = await request.json()
    request.app.logger.info(params)
    response = await call_next(request)
    return response


import uvicorn

uvicorn.run(app, port=8000)
```

And then:
curl --location --request POST 'http://localhost:8000/' --data-raw '{"a": 1}'

**Comment by panla** [comment]
@juntatalor

thanks!

I have a simple demo "https://github.com/panla/fastapi_gino_pgsql"
 
need database.

middleware in "https://github.com/panla/fastapi_gino_pgsql/blob/master/apps/middleware/request.py"
line 19 and 20

curl --location --request POST 'http://localhost:8000/api/v1/users' --data-raw '{"name": "22"}'
will block or I have bug in code

**Comment by panla** [comment]
### 处理方案

参考 **https://github.com/tiangolo/fastapi/issues/394**

thanks `@ruslankrivoshein`

**Comment by panla** [comment]
issue 394

**Comment by tiangolo** [comment]
Thanks for the help here @juntatalor ! 👏  🙇 

Thanks for reporting back and closing the issue @panla  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
